daemon/list.go

```diff
@@ -174,70 +203,39 @@ func (daemon *Daemon) filterByNameIDMatches(view container.View, ctx *listContex
 	return cntrs, nil
 }
 
-// reduceContainers parses the user's filtering options and generates the list of containers to return based on a reducer.
-func (daemon *Daemon) reduceContainers(config *types.ContainerListOptions, reducer containerReducer) ([]*types.Container, error) {
+// rangeContainers parses the user's filtering options and ranges over the list of matching containers.
+// If f returns false, range stops the iteration.
+func (daemon *Daemon) rangeContainers(config *types.ContainerListOptions, f func(*container.Snapshot, *listContext) bool) error {
 	if err := config.Filters.Validate(acceptedPsFilterTags); err != nil {
-		return nil, err
+		return err
 	}
 
-	var (
-		view       = daemon.containersReplica.Snapshot()
-		containers = []*types.Container{}
-	)
-
+	view := daemon.containersReplica.Snapshot()
 	ctx, err := daemon.foldFilter(view, config)
 	if err != nil {
-		return nil, err
+		return err
 	}
 
 	// fastpath to only look at a subset of containers if specific name
 	// or ID matches were provided by the user--otherwise we potentially
 	// end up querying many more containers than intended
 	containerList, err := daemon.filterByNameIDMatches(view, ctx)
 	if err != nil {
-		return nil, err
+		return err
 	}
 
-	for i := range containerList {
-		t, err := daemon.reducePsContainer(&containerList[i], ctx, reducer)
-		if err != nil {
-			if err != errStopIteration {
-				return nil, err
-			}
-			break
+	for _, cntr := range containerList {
+		switch includeContainerInList(&cntr, ctx) {
+		case excludeContainer:
+			continue
+		case stopIteration:
+			return nil
 		}
-		if t != nil {
-			containers = append(containers, t)
-			ctx.idx++
+		if !f(&cntr, ctx) {
+			return nil
 		}
 	}
-
-	return containers, nil
-}
-
-// reducePsContainer is the basic representation for a container as expected by the ps command.
-func (daemon *Daemon) reducePsContainer(container *container.Snapshot, ctx *listContext, reducer containerReducer) (*types.Container, error) {
-	// filter containers to return
-	switch includeContainerInList(container, ctx) {
-	case excludeContainer:
-		return nil, nil
-	case stopIteration:
-		return nil, errStopIteration
-	}
-
-	// transform internal container struct into api structs
-	newC, err := reducer(container, ctx)
-	if err != nil {
-		return nil, err
-	}
-
-	// release lock because size calculation is slow
```

Comment:
Not sure what this is referring to, since there are no locks here; removed the comment